src/Compilers/CSharp/Test/Semantic/Semantics/UsingStatementTests.cs

```diff
@@ -305,14 +403,396 @@ static void Main()
+                Diagnostic(ErrorCode.WRN_PatternBadSignature, "C1 c = new C1()").WithArguments("C1", "disposable", "N3.C4.Dispose(C1)").WithLocation(64, 20),
+                // (64,20): error CS1674: 'C1': type used in a using statement must be implicitly convertible to 'System.IDisposable' or have a public void-returning Dispose() instance method.
+                //             using (C1 c = new C1()) // error 2: C4.Dispose does not match pattern
+                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "C1 c = new C1()").WithArguments("C1").WithLocation(64, 20),
+                // (78,20): warning CS0278: 'C1' does not implement the 'disposable' pattern. 'C2.Dispose(C1)' is ambiguous with 'C4.Dispose(C1)'.
+                //             using (C1 c = new C1())  // error 3: C2.Dispose and C4.Dispose are ambiguous
+                Diagnostic(ErrorCode.WRN_PatternIsAmbiguous, "C1 c = new C1()").WithArguments("C1", "disposable", "N1.C2.Dispose(C1)", "N3.C4.Dispose(C1)").WithLocation(78, 20),
+                // (78,20): error CS1674: 'C1': type used in a using statement must be implicitly convertible to 'System.IDisposable' or have a public void-returning Dispose() instance method.
+                //             using (C1 c = new C1())  // error 3: C2.Dispose and C4.Dispose are ambiguous
+                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "C1 c = new C1()").WithArguments("C1").WithLocation(78, 20),
+                // (93,24): warning CS0280: 'C1' does not implement the 'disposable' pattern. 'C4.Dispose(C1)' has the wrong signature.
+                //                 using (C1 c = new C1())  // error 4: C4.Dispose does not match pattern
+                Diagnostic(ErrorCode.WRN_PatternBadSignature, "C1 c = new C1()").WithArguments("C1", "disposable", "N3.C4.Dispose(C1)").WithLocation(93, 24),
+                // (93,24): error CS1674: 'C1': type used in a using statement must be implicitly convertible to 'System.IDisposable' or have a public void-returning Dispose() instance method.
+                //                 using (C1 c = new C1())  // error 4: C4.Dispose does not match pattern
+                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "C1 c = new C1()").WithArguments("C1").WithLocation(93, 24)
+                );
+        }
+
+        [Fact]
+        public void UsingPatternWithMultipleExtensionTargets()
+        {
+            var source = @"
+class C1
+{
+}
+
+class C2
+{
+}
+
+static class C3 
+{
+    public static void Dispose(this C1 c1) { }
+
+    public static void Dispose(this C2 c2) { }
+
+}
+
+class C4
+{
+    static void Main()
+    {
+        using (C1 c = new C1())
+        {
+        }
+        C1 c1b = new C1();
+        using (c1b) { }
+
+        using (C2 c = new C2())
+        {
+        }
+        C2 c2b = new C2();
+        using (c2b) { }
+    }
+}";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void UsingPatternWithLessDerivedTarget()
```

Comment:
>UsingPatternWithLessDerivedTarget [](start = 20, length = 33)

Please add a test with `Dispose` extension methods for both `C1` and `C2` to verify overload resolution when calling with an instance of each.

Reply:
Please also test a mix of generic and non-generic methods:
```
static void Dispose(this C1 c) { }
static void Dispose<T>(this T t) where T : C2 { }
```

---
In reply to: [221748751](https://github.com/dotnet/roslyn/pull/29307#discussion_r221748751) [](ancestors = 221748751)

Reply:
Other variants:

```
static void Dispose(in SomeStruct) 
static void Dispose(ref SomeStruct)
```

---
In reply to: [221754221](https://github.com/dotnet/roslyn/pull/29307#discussion_r221754221) [](ancestors = 221754221,221748751)